File: daily/rust/Dec-17-Design-a-food-system.rs
struct Food {
    food_rating: i32,
    food_name: String,
}

impl Food {
    fn new(food_rating: i32, food_name: String) -> Self {
        Self {
            food_rating,
            food_name,
        }
    }

    // compare food_rating
    fn cmp(&self, other: &Self) -> bool {
        if (self.food_rating == other.food_rating) {
            return self.food_name > other.food_name;
        }

        self.food_rating < other.food_rating
    }
}

struct FoodRatings {
    // map food w/rating
    food_rating_map: HashMap<String, i32>,
    // map food w/cuisine type
    food_cuisine_map: HashMap<String, String>,

    // map cuisine w/food (sort by rating/name)
    // vector element -> Food: (food_rating, food_name)
    cuisine_food_map: HashMap<String, Vec<(i32, String)>>,
}


/**
 * `&self` means the method takes an immutable reference.
 * if you need a mutable reference, change it to `&mut self` instead.
 */
impl FoodRatings {

    fn new(foods: Vec<String>, cuisines: Vec<String>, ratings: Vec<i32>) -> Self {
        let mut food_rating_map = HashMap::new();
        let mut food_cuisine_map = HashMap::new();
        let mut cuisine_food_map = HashMap::new();

        for i in 0..foods.len() {
            // store 'rating' 'cuisine' of food in food_rating_map, food_cuisine_map
            food_rating_map.insert(foods[i].clone(), ratings[i]);
            food_cuisine_map.insert(foods[i].clone(), cuisines[i].clone());
            // insert (rating, name) in cuisine_food_map
            let cuisine = cuisines[i].clone();
            cuisine_food_map.entry(cuisine).or_insert(Vec::new()).push((ratings[i], foods[i].clone()));
        }

    }

    fn change_rating(&self, food: String, new_rating: i32) {
        // update food's rating in food_rating_map
        food_rating_map.insert(food, new_rating);
        // update food's rating in cuisine_food_map
        cuisine_food_map.get_mut(food_cuisine_map.get(food)).unwrap().sort_by(|a, b| a.cmp(b));
    }

    fn highest_rated(&self, cuisine: String) -> String {
        // get highest rated food in cuisine_food_map
        let mut highest_rated = cuisine_food_map.get(cuisine).unwrap().last().unwrap().1;

        // if latest rating of food doesn't match with ratings in food vector
        // discard this food
        while food_rating_map.get(highest_rated.food_name) != highest_rated.food_rating {
            cuisine_food_map.get_mut(cuisine).unwrap().pop();
            highest_rated = cuisine_food_map.get(cuisine).unwrap().last().unwrap().1;
        }
        highest_rated.food_name
    }
}

File: daily/cpp/Dec-17-Design-a-food-system.cpp
#include <queue>
#include <string>
#include <unordered_map>
using namespace std;

class Food {
 public:
  // store food's rating
  int foodRating;
  // food's name
  string foodName;

  Food(int foodRating, string foodName) {
    this->foodRating = foodRating;
    this->foodName = foodName;
  }

  // overload operator < for comparison
  bool operator<(const Food& other) const {
    // food ratings are same sort by name
    // lexographically smaller name food will be on top
    if (foodRating == other.foodRating) return foodName > other.foodName;

    // sort on the basis of food rating
    // (bigger rating food on top)
    return foodRating < other.foodRating;
  }
};

class FoodRatings {
 private:
  // map food w/ its rating
  unordered_map<string, int> foodRatingMap;
  // map food w/ cuisine type
  unordered_map<string, string> foodCuisineMap;

  // store all food of a cuisine in priority queue (sorted by ratings/name)
  // Priority queue element -> Food: (foodRating, foodName)
  unordered_map<string, priority_queue<Food>> cuisineFoodMap;

 public:
  FoodRatings(vector<string>& foods, vector<string>& cuisines,
              vector<int>& ratings) {
    for (int i = 0; i < foods.size(); ++i) {
      // store 'rating' and 'cuisine' of current 'food' in 'foodRatingMap' and
      // 'foodCuisineMap' maps
      foodRatingMap[foods[i]] = ratings[i];
      foodCuisineMap[foods[i]] = cuisines[i];
      // insert the '(rating, food)' pair in current cuisine's priority queue
      cuisineFoodMap[cuisines[i]].push(Food(ratings[i], foods[i]));
    }
  }

  void changeRating(string food, int newRating) {
    // update food's rating in 'foodRatingMap'
    foodRatingMap[food] = newRating;
    // insert the '(new rating, name)' element in respective cuisine's priority
    // queue
    auto cuisineName = foodCuisineMap[food];
    cuisineFoodMap[cuisineName].push(Food(newRating, food));
  }

  string highestRated(string cuisine) {
    // get the highest rated 'food' from cuisine's priority queue
    auto highestRated = cuisineFoodMap[cuisine].top();

    // if latest rating of 'food' doesn't match with 'rating' on which it was
    // stored in priority queue discard this element of the priority queue
    while (foodRatingMap[highestRated.foodName] != highestRated.foodRating) {
      cuisineFoodMap[cuisine].pop();
      highestRated = cuisineFoodMap[cuisine].top();
    }

    // return the highest rated food
    return highestRated.foodName;
  }
};